lib/IRGen/IRGenSIL.cpp

```diff
@@ -2472,6 +2476,53 @@ static void emitDynamicSelfMetadata(IRGenSILFunction &IGF) {
   IGF.setDynamicSelfMetadata(selfTy, isExact, value, selfKind);
 }
 
+/// C++ interop might refer to the type metadata in some scenarios.
+/// This function covers those cases and makes sure metadata is emitted
+/// for the foreign types.
+static void noteUseOfMetadataByCXXInterop(IRGenerator &IRGen,
+                                          const SILFunction *f,
+                                          const TypeExpansionContext &context) {
+  auto type = f->getLoweredFunctionType();
+
+  // Notes the use of foreign types in generic arguments for C++ interop.
+  auto processType = [&](CanType type) {
+    struct Walker : TypeWalker {
+      Walker(IRGenerator &IRGen) : IRGen(IRGen) {}
+
+      Action walkToTypePre(Type ty) override {
+        if (auto *BGT = ty->getAs<BoundGenericType>())
+          genericDepth++;
+        else if (auto *nominal = ty->getAs<NominalType>())
```

Comment:
I'm not sure I understand the goal of `genericDepth` here. It's going to increase whenever we're walking into a type like `X<Y>`, and suppress the type metadata for both `X` and `Y`. Is that intended?

My expectation based on the comment is that you wanted to emit metadata for any non-generic types encountered along the way. To do that, you could skip the `genericDepth` bit and instead check whether the nominal type declaration has a generic signature where `areAllParamsConcrete()` is false.

Reply:
My goal is to emit metadata for the foreign type `Y` when it was a generic argument, e.g., in `X<Y>`. 

My understanding is that we visit `X` first, increase the counter, visit `Y` later with the counter set to 1, so we emit metadata for Y. I only wanted to avoid emitting metadata for the outermost types (in this case `X`). 

> My expectation based on the comment is that you wanted to emit metadata for any non-generic types encountered along the way.

Emitting metadata for every non-generic type would emit metadata for `X` (a single non-generic type). But we do not need metadata for that, only for types that are used as generic arguments.

Reply:
Okay! I misunderstood the premise. Looks good, then.

Reply:
Hmm, I think the most probably the most concise way to describe what I wanted to do here is to collect all the generic arguments recursively.